### PR TITLE
CASSANDRA-17217 Add proper histogram checkup to fix flaky failThresholdSinglePartition

### PR DESCRIPTION
**Pull request by bbotella**
Add proper hitogram checkup to fix flaky failThresholdSinglePartition test
patch by Bernardo Botella Corbi

### CODE REVIEW

File: test/distributed/org/apache/cassandra/distributed/test/trackwarnings/AbstractClientSizeWarning.java [hunk 1]
```diff
@@ -283,15 +295,26 @@ public boolean matches(InetAddress value)
...
+        ICoordinator node = CLUSTER.coordinator(1);
+        for (int i = 0; i < failThresholdRowCount(); i++)
+            node.execute("INSERT INTO " + KEYSPACE + ".tbl (pk, ck, v) VALUES (1, ?, ?)", ConsistencyLevel.ALL, i + 1, bytes(512));
+
+        if (shouldFlush())
+            CLUSTER.stream().forEach(i -> i.flush(KEYSPACE));
 
         // query should no longer fail
         enable(false);
+        checkpointHistogram();
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/trackwarnings/AbstractClientSizeWarning.java:311]: this has to go after the query

Comment (bbotella) [reply test/distributed/org/apache/cassandra/distributed/test/trackwarnings/AbstractClientSizeWarning.java:311]: I was thinking that it is maybe better to remove this checkpointHistogram completely from here, and just not change what current test is doing
https://github.com/apache/cassandra/blob/trunk/test/distributed/org/apache/cassandra/distributed/test/trackwarnings/AbstractClientSizeWarning.java#L195

File: CHANGES.txt [hunk 2]
```diff
@@ -1,3 +1,5 @@
+Add proper hitogram checkup to fix flaky failThresholdSinglePartition test
```
Comment (dcapwell) [inline CHANGES.txt]: not proper format, I would personally remove and leave to committer

If you want it correctly (which always conflicts on merge so committer has to fix... hence why I always exclude... but thats me), you would put this under `4.1` (if its going in the 4.1 release), put it at the top and add ` * ` prefix

File: test/distributed/org/apache/cassandra/distributed/test/trackwarnings/AbstractClientSizeWarning.java [hunk 3]
```diff
@@ -283,15 +295,26 @@ public boolean matches(InetAddress value)
...
+        for (int i = 0; i < failThresholdRowCount(); i++)
+            node.execute("INSERT INTO " + KEYSPACE + ".tbl (pk, ck, v) VALUES (1, ?, ?)", ConsistencyLevel.ALL, i + 1, bytes(512));
+
+        if (shouldFlush())
+            CLUSTER.stream().forEach(i -> i.flush(KEYSPACE));
 
         // query should no longer fail
         enable(false);
         SimpleQueryResult result = node.executeWithResult(cql, ConsistencyLevel.ALL);
+        checkpointHistogram();
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/trackwarnings/AbstractClientSizeWarning.java:311]: this has to be before query, as we are checking if the query triggered the condition or not

File: CHANGES.txt (hunk 2 above)
Comment (yifan-c) [reply CHANGES.txt]: > not proper format, I would personally remove and leave to committer

Agree. 

And... this patch is to fix test only. So we will not add an entry in the `CHANGES.txt`.

File: test/distributed/org/apache/cassandra/distributed/test/trackwarnings/AbstractClientSizeWarning.java [hunk 4]
```diff
@@ -283,15 +295,26 @@ public boolean matches(InetAddress value)
...
+
+        if (shouldFlush())
+            CLUSTER.stream().forEach(i -> i.flush(KEYSPACE));
 
         // query should no longer fail
         enable(false);
+        checkpointHistogram();
         SimpleQueryResult result = node.executeWithResult(cql, ConsistencyLevel.ALL);
         assertThat(result.warnings()).isEmpty();
         assertHistogramNotUpdated();
```
Comment (sathyakplm) [inline test/distributed/org/apache/cassandra/distributed/test/trackwarnings/AbstractClientSizeWarning.java:314]: One minor query. In Line 313, we are asserting that the warnings is empty. Isn't that sufficient? Do we still need to check if the histogram hasn't updated? Isn't it redundant?

Comment (bbotella) [reply test/distributed/org/apache/cassandra/distributed/test/trackwarnings/AbstractClientSizeWarning.java:314]: I didn't want to change current behavior. As you can see in the original test:
https://github.com/apache/cassandra/blob/trunk/test/distributed/org/apache/cassandra/distributed/test/trackwarnings/AbstractClientSizeWarning.java#L206-L207

It was asserting both things already. Same goes for the failThreshold method:
https://github.com/apache/cassandra/blob/trunk/test/distributed/org/apache/cassandra/distributed/test/trackwarnings/AbstractClientSizeWarning.java#L290-L291

**Pull request closed without merging**